source/common/ratelimit/ratelimit_impl.cc

```diff
@@ -46,39 +50,55 @@ void GrpcClientImpl::createRequest(pb::lyft::ratelimit::RateLimitRequest& reques
 void GrpcClientImpl::limit(RequestCallbacks& callbacks, const std::string& domain,
                            const std::vector<Descriptor>& descriptors,
                            const Tracing::TransportContext& context) {
-  ASSERT(!callbacks_);
+  ASSERT(callbacks_ == nullptr);
   callbacks_ = &callbacks;
   context_ = context;
 
   pb::lyft::ratelimit::RateLimitRequest request;
   createRequest(request, domain, descriptors);
-  service_.ShouldRateLimit(nullptr, &request, &response_, nullptr);
+
+  stream_ = async_client_->start(service_method_, *this, timeout_);
+  if (stream_ == nullptr) {
+    callbacks_->complete(LimitStatus::Error);
+    callbacks_ = nullptr;
+    return;
+  }
+  stream_->sendMessage(request);
 }
 
-void GrpcClientImpl::onPreRequestCustomizeHeaders(Http::HeaderMap& headers) {
+void GrpcClientImpl::onCreateInitialMetadata(Http::HeaderMap& metadata) {
   if (!context_.request_id_.empty()) {
-    headers.insertRequestId().value(context_.request_id_);
+    metadata.insertRequestId().value(context_.request_id_);
   }
 
   if (!context_.span_context_.empty()) {
-    headers.insertOtSpanContext().value(context_.span_context_);
+    metadata.insertOtSpanContext().value(context_.span_context_);
   }
 }
 
-void GrpcClientImpl::onSuccess() {
-  LimitStatus status = LimitStatus::OK;
-  ASSERT(response_.overall_code() != pb::lyft::ratelimit::RateLimitResponse_Code_UNKNOWN);
-  if (response_.overall_code() == pb::lyft::ratelimit::RateLimitResponse_Code_OVER_LIMIT) {
-    status = LimitStatus::OverLimit;
+void GrpcClientImpl::onReceiveMessage(
+    std::unique_ptr<pb::lyft::ratelimit::RateLimitResponse>&& message) {
+  response_ = std::move(message);
+}
+
+void GrpcClientImpl::onRemoteClose(Grpc::Status::GrpcStatus status) {
+  if (status != Grpc::Status::GrpcStatus::Ok) {
+    callbacks_->complete(LimitStatus::Error);
+    callbacks_ = nullptr;
+    stream_ = nullptr;
+    return;
   }
 
-  callbacks_->complete(status);
-  callbacks_ = nullptr;
-}
+  LimitStatus limit_status = LimitStatus::OK;
```

Comment:
Question: It seems kind of odd to do this logic in the context of onRemoteClose(). Why not onReceiveMessage() ? I guess the larger question is for unary APIs (which are very common) can we make coding against this easier vs. having to remember to close the stream, etc.?